ccl/backupccl: TestDataDriven_user_defined_functions_in_checks failed

ccl/backupccl.TestDataDriven_user_defined_functions_in_checks [failed](https://tanzanite.cluster.engflow.com/invocations/default/b8827329-ddca-46a9-9b31-a41b550a4bfb?testReportRun=12&testReportShard=26&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on release-24.1 @ [5d952f80b3e1efe2e9aaed73f1fd68433880fcb7](https://github.com/cockroachdb/cockroach/commits/5d952f80b3e1efe2e9aaed73f1fd68433880fcb7):


```
    datadriven.go:144: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks:143:
        exec-sql [1 args]
        CREATE DATABASE db1;
        CREATE DATABASE db2;
        CREATE DATABASE db3;
        USE db1;
        CREATE SCHEMA sc1;
        CREATE FUNCTION sc1.f1(a INT) RETURNS INT LANGUAGE SQL AS $$
          SELECT a + 1;
        $$;
        CREATE TABLE sc1.t1(a INT PRIMARY KEY, b INT CHECK(sc1.f1(b) > 1));
        ----
    datadriven.go:144: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks:155:
        exec-sql [0 args]
        BACKUP DATABASE db1 INTO 'nodelocal://1/test/'
        ----
    datadriven.go:144: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks:159:
        query-sql [0 args]
        WITH descs AS (
          SHOW BACKUP LATEST IN 'nodelocal://1/test/'
        )
        SELECT database_name, parent_schema_name, object_name, object_type, is_full_cluster FROM descs
        ----
        <nil> <nil> db1 database false
        db1 <nil> public schema false
        db1 <nil> sc1 schema false
        db1 sc1 f1 function false
        db1 sc1 t1 table false
    datadriven.go:144: 
        /var/lib/engflow/worker/work/1/exec/bazel-out/k8-fastbuild/bin/pkg/ccl/backupccl/backupccl_test_/backupccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks:171:
         RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db2';
        expected:
        pq: cannot restore table "t1" without referenced function 114 (or "skip_missing_udfs" option)
        
        found:
        pq: cannot restore table "t1" without referenced function 116 (or "skip_missing_udfs" option)
    testutils.go:289: no Invalid Descriptors
    testutils.go:289: no Invalid Descriptors
    testutils.go:289: no Invalid Descriptors
    testutils.go:289: no Invalid Descriptors
    testutils.go:266: Warning: Could not check for invalid descriptors: sql: database is closed
    testutils.go:266: Warning: Could not check for invalid descriptors: sql: database is closed
    testutils.go:266: Warning: Could not check for invalid descriptors: sql: database is closed
    testutils.go:266: Warning: Could not check for invalid descriptors: sql: database is closed
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestDataDriven_user_defined_functions_in_checks3963945615
--- FAIL: TestDataDriven_user_defined_functions_in_checks (17.80s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=12</code>
 - <code>shard=26</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121102 ccl/backupccl: TestDataDriven_user_defined_functions_in_checks failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDataDriven_user_defined_functions_in_checks.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/backupccl/testdata/backup-restore/plpgsql_procedures
+++ b/pkg/ccl/backupccl/testdata/backup-restore/plpgsql_procedures
@@ -379,9 +379,10 @@ WITH to_json AS (
       system.descriptor
     WHERE id = $defaultdb_sc1_id
 )
-SELECT d->'schema'->>'functions'::string FROM to_json;
+-- Remove ID from the result, since it isn't stable.
+SELECT json_set(d, '{schema,functions,p,signatures,0,id}', '0')->'schema'->>'functions'::string FROM to_json;
 ----
-{"p": {"signatures": [{"id": 111, "isProcedure": true, "returnType": {"family": "VoidFamily", "oid": 2278}}]}}
+{"p": {"signatures": [{"id": 0, "isProcedure": true, "returnType": {"family": "VoidFamily", "oid": 2278}}]}}
 
 exec-sql
 BACKUP TABLE sc1.t INTO 'nodelocal://1/test/'

--- a/pkg/ccl/backupccl/testdata/backup-restore/plpgsql_user_defined_functions
+++ b/pkg/ccl/backupccl/testdata/backup-restore/plpgsql_user_defined_functions
@@ -405,9 +405,10 @@ WITH to_json AS (
       system.descriptor
     WHERE id = $defaultdb_sc1_db
 )
-SELECT d->'schema'->>'functions'::string FROM to_json;
+-- Remove ID from the result, since it isn't stable.
+SELECT json_set(d, '{schema,functions,f,signatures,0,id}', '0')->'schema'->>'functions'::string FROM to_json;
 ----
-{"f": {"signatures": [{"id": 111, "returnType": {"family": "IntFamily", "oid": 20, "width": 64}}]}}
+{"f": {"signatures": [{"id": 0, "returnType": {"family": "IntFamily", "oid": 20, "width": 64}}]}}
 
 exec-sql
 BACKUP TABLE sc1.t INTO 'nodelocal://1/test/'

--- a/pkg/ccl/backupccl/testdata/backup-restore/procedures
+++ b/pkg/ccl/backupccl/testdata/backup-restore/procedures
@@ -351,9 +351,10 @@ WITH to_json AS (
       system.descriptor
     WHERE id = $defaultdb_sc1_db
 )
-SELECT d->'schema'->>'functions'::string FROM to_json;
+-- Remove ID from the result, since it isn't stable.
+SELECT json_set(d, '{schema,functions,p,signatures,0,id}', '0')->'schema'->>'functions'::string FROM to_json;
 ----
-{"p": {"signatures": [{"id": 111, "isProcedure": true, "returnType": {"family": "VoidFamily", "oid": 2278}}]}}
+{"p": {"signatures": [{"id": 0, "isProcedure": true, "returnType": {"family": "VoidFamily", "oid": 2278}}]}}
 
 exec-sql
 BACKUP TABLE sc1.t INTO 'nodelocal://1/test/'

--- a/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions
+++ b/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions
@@ -347,9 +347,10 @@ WITH to_json AS (
       system.descriptor
     WHERE id = $defaultdb_sc1_db
 )
-SELECT d->'schema'->>'functions'::string FROM to_json;
+-- Remove ID from the result, since it isn't stable.
+SELECT json_set(d, '{schema,functions,f,signatures,0,id}', '0')->'schema'->>'functions'::string FROM to_json;
 ----
-{"f": {"signatures": [{"id": 111, "returnType": {"family": "IntFamily", "oid": 20, "width": 64}}]}}
+{"f": {"signatures": [{"id": 0, "returnType": {"family": "IntFamily", "oid": 20, "width": 64}}]}}
 
 exec-sql
 BACKUP TABLE sc1.t INTO 'nodelocal://1/test/'

--- a/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks
+++ b/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-checks
@@ -168,10 +168,10 @@ db1 <nil> sc1 schema false
 db1 sc1 f1 function false
 db1 sc1 t1 table false
 
-exec-sql
+exec-sql expect-error-regex=(cannot restore table "t1" without referenced function [0-9]+ \(or "skip_missing_udfs" option\))
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db2';
 ----
-pq: cannot restore table "t1" without referenced function 114 (or "skip_missing_udfs" option)
+regex matches error
 
 exec-sql
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db2', skip_missing_udfs;
@@ -209,10 +209,10 @@ SELECT database_name, parent_schema_name, object_name, object_type, is_full_clus
 db1 <nil> sc1 schema false
 db1 sc1 t1 table false
 
-exec-sql
+exec-sql expect-error-regex=(cannot restore table "t1" without referenced function [0-9]+ \(or "skip_missing_udfs" option\))
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db3';
 ----
-pq: cannot restore table "t1" without referenced function 114 (or "skip_missing_udfs" option)
+regex matches error
 
 exec-sql
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db3', skip_missing_udfs;

--- a/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-defaults
+++ b/pkg/ccl/backupccl/testdata/backup-restore/user-defined-functions-in-defaults
@@ -170,10 +170,10 @@ db1 <nil> sc1 schema false
 db1 sc1 f1 function false
 db1 sc1 t1 table false
 
-exec-sql
+exec-sql expect-error-regex=(cannot restore table "t1" without referenced function [0-9]+ \(or "skip_missing_udfs" option\))
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db2';
 ----
-pq: cannot restore table "t1" without referenced function 114 (or "skip_missing_udfs" option)
+regex matches error
 
 exec-sql
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db2', skip_missing_udfs;
@@ -211,10 +211,10 @@ SELECT database_name, parent_schema_name, object_name, object_type, is_full_clus
 db1 <nil> sc1 schema false
 db1 sc1 t1 table false
 
-exec-sql
+exec-sql expect-error-regex=(cannot restore table "t1" without referenced function [0-9]+ \(or "skip_missing_udfs" option\))
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db3';
 ----
-pq: cannot restore table "t1" without referenced function 114 (or "skip_missing_udfs" option)
+regex matches error
 
 exec-sql
 RESTORE TABLE sc1.t1 FROM LATEST IN 'nodelocal://1/test/' WITH into_db = 'db3', skip_missing_udfs;

